azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,441 @@
+            topic=constant.PUBLISH_TOPIC_REGISTRATION.format(rid),
+            request=" ",
+            callback=self._handle_register_response_received,
+        )
+
+    def _query_operation_status(self, event_data):
+        """
+        Poll the service for operation status.
+        """
+        logger.info("Querying operation status from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+        result = event_data.args[0].args[0]
+
+        operation_id = result.operation_id
+        self._operations[rid] = constant.PUBLISH_TOPIC_QUERYING.format(rid, operation_id)
+        self._request_response_provider.send_request(
+            rid=rid,
+            topic=constant.PUBLISH_TOPIC_QUERYING.format(rid, operation_id),
+            request=" ",
+            callback=self._handle_query_response_received,
+        )
+
+    def _handle_register_response_received(self, url_portion, key_values_dict, response):
+        """
+        The function to call in case of a response from a registration request.
+        :param url_portion: The portion of the url containing the status code.
+        :param key_values_dict: The dictionary containing the query parameters of the returned topic.
+        :param response: The complete response from the service.
+        """
+        self._query_timer.cancel()
+
+        url_parts = url_portion.split("/")
+        status_code = url_parts[POS_STATUS_CODE_IN_TOPIC]
+        rid = str(key_values_dict["rid"][0])
+        retry_after = (
+            None if "retry-after" not in key_values_dict else str(key_values_dict["retry-after"][0])
+        )
+        intermediate_registration_result = RegistrationQueryStatusResult(rid, retry_after)
+
+        if int(status_code, 10) >= 429:
+            del self._operations[rid]
+            self._trig_wait(intermediate_registration_result)
+        elif int(status_code, 10) >= 300:  # pure failure
+            self._trig_error(ValueError("Incoming message failure"))  # TODO:Force status=error?
+        else:  # successful case, transition into complete or poll status
+            self._handle_successful_response(rid, retry_after, response)
+
+    def _handle_query_response_received(self, url_portion, key_values_dict, response):
+        """
+        The function to call in case of a response from a polling/query request.
+        :param url_portion: The portion of the url containing the status code.
+        :param key_values_dict: The dictionary containing the query parameters of the returned topic.
+        :param response: The complete response from the service.
+        """
+        self._query_timer.cancel()
+        self._polling_timer.cancel()
+
+        url_parts = url_portion.split("/")
+        status_code = url_parts[POS_STATUS_CODE_IN_TOPIC]
+        rid = str(key_values_dict["rid"][0])
+        retry_after = (
+            None if "retry-after" not in key_values_dict else str(key_values_dict["retry-after"][0])
+        )
```

Comment:
Again, as above, a lot of this url stuff should be abstracted to another module

Reply:
I will try moving them to another module. Once I do pipeline

---
In reply to: [282968098](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282968098) [](ancestors = 282968098)